[WIP] Split SR into QGTMatrix and solver. (#674)
netket/netket: docs/docs/sr.md

A complication is given by the fact that the QGT is determined by Monte Carlo sampling and it might have several eigenvalues that are zero or very small, leading to numerical stability issues when inverting the matrix or in the resulting dynamics.
The linear system can be solved with several methods. For the models with many parameters and to achieve the best performance, iterative solvers such as those found in `jax.scipy.sparse.linalg`, such as {ref}`jax.scipy.sparse.linalg.cg` {ref}`jax.scipy.sparse.linalg.gmres` are the best choice. 
Do note that to stabilize those algorithms it is often needed to add a small ($10^{-5} - 10^{-2}$) shift to the diagonal of the QGT. 
This can be set with the keyword argument `diag_shift`.
Those methods, combined with our lazy representations of the QGT, never instantiate the full matrix and therefore achieve a great performance.
However, when the number of parameters is small (smaller than 1000-5000), it might make sense to solve the system by factorizing the QGT with cholesky or SVD.
Those techniques require instantiating the full dense matrix, but in general are more stable and don't require a diagonal shift.

> ### Using the stochastic reconfiguration
> 
> To use the Stochastic reconfiguration, you must simply provide it as a preconditioner to the VMC solver.
> 

Reviewer: ```suggestion
### Using stochastic reconfiguration

To use SR, you must simply provide it as a preconditioner to the VMC solver.

```

# Quantum Geometric Tensor ( and Stochastic Reconfiguration )

The Quantum Geometric Tensor (QGT) is the Fubini-Study metric tensor of the manifold on which a variational state is defined.

> In general it is not possible to compute a form of the QGT valid on the whole manifold, but we can compute local approximations valid in small region around a given point $ W $

Reviewer: I think this is mathematically imprecise? The metric on a manifold is always defined locally (of course it can happen to be the same at each point in a flat geometry), so making it sound like we are approximating something (a "exact" QGT that would be valid on the whole manifold) is a bit misleading since such an object just doesn't exists (so there is nothing to "approximate"). We _are_ computing the QGT at the specific point _W_; the approximation only happens when we do VMC.
```suggestion
In the general case (and in practise for any NQS ansatz), the QGT varies depending on the current quantum state and thus needs to be computed when the variational parameters $W$ change.
```

Author: Yes you are right.
However we are still computing the linearization of the metric tensor on a point W , not the actual metric tensor, which is an approximation.

Reviewer: Sure? I think the [metric tensor](https://en.wikipedia.org/wiki/Metric_tensor) on a manifold includes this linearization by definition. Note how the Wikipedia article says:
> Equipped with this notion of length [= length of a geodesic between two points], a Riemannian manifold is a metric space, meaning that it has a distance function d(p, q) whose value at a pair of points p and q is the distance from p to q. Conversely, the metric tensor itself is the derivative of the distance function (taken in a suitable manner). Thus the metric tensor gives the infinitesimal distance on the manifold.

So the metric tensor already gives the infinitesimal distance around a given points, which is what you are saying, right? But in that sense it's really not an approximation. It does locally approximate the _metric_ (in the sense of distance function), but that's a different object.

(Just to be clear, I'm not arguing this to be pedantic - we already agree on the formulation in the docs I think. I'd just like to understand whether we use the same terminology or not. :slightly_smiling_face: )

To give an example, we consider a variational ansatz $ \psi\colon \mathbb{R} \rightarrow \mathscr{H} $ which maps elements $W$ of the parameter space to vectors in Hilbert space (quantum states) $\psi_W$.

> The parameter space has a simple Euclidean metric, therefore the distance between two points $\Vert \bf{W} \Vert$ and $\bf{W} + \bf{\delta W} $ is simply  $ \bf{\delta W} $.

Reviewer: ```suggestion
The parameter space has a simple Euclidean metric, therefore the distance between two points $\bf{W}$ and $\bf{W} + \bf{\delta W}$ is simply  $\Vert\bf{\delta W} \Vert$.
```
Oops, I put the `‖ ‖`s in the wrong place.

> To give an example, we consider a variational wavefunction $ \psi : \mathbb{R} \rightarrow \mathscr{H} $ which maps elements of the variational manifold $ W $ to vectors in the Hilbert space (wavefunctions) $ \psi_W $.
> 
> The variational manifold has a simple Euclidean metric, therefore the distance between two points $ \bf{W} $ and $\bf{W} + \bf{\delta W} $ is simply  $ \bf{\delta W} $.
> However, what we really are interested in when optimizing variational ansatze is not the Euclidean distance between those two poins, but rather the distance in the Hilbert space between the elements in the co-domain.
> As the objects $\psi_{\bf{W}}$ is defined on an Hilbert space, the correct metric there is $ L_2 $ and not the Euclidean norm, we consider the distance defined as

Reviewer: You mean parameter space when you say "variational manifold" sometimes, right? The _variational manifold_ is almost everywhere defined as $M = \{\psi_W | W \in C^M \}$ and is thus part of the Hilbert space and definitely not Euclidean.
```suggestion
To give an example, we consider a variational ansatz $ \psi\colon \mathbb{R} \rightarrow \mathscr{H} $ which maps elements $W$ of the parameter space to vectors in Hilbert space (quantum states) $\psi_W$.

The parameter space has a simple Euclidean metric, therefore the distance between two points $ \bf{W} $ and $\bf{W} + \bf{\delta W} $ is simply  $ \bf{\delta W} $.
However, what we really are interested in when optimizing variational ansätze is not the (Euclidean) distance between those two points in parameter space, but rather the distance in the Hilbert space between the corresponding quantum states (which also properly takes into account gauge degrees of freedom).
As the objects $\psi_{\bf{W}}$ is defined on an Hilbert space, the correct metric there is $ L_2 $ and not the Euclidean norm, we consider the distance defined as
```
The last paragraph seems incomplete, so I've left it as is for now.

Author: Yeah i'm still writing this...
Yes. I meant parameter space, thanks for catching this.

This is the Quantum Geometric Tensor.

> In NetKet you cna obtain (a local approximation of) the quantum geometric tensor of a variational state by calling `variational_state.quantum_geometric_tensor())`.

Reviewer 2: ```suggestion
In NetKet you can obtain (a local approximation of) the quantum geometric tensor of a variational state by calling `variational_state.quantum_geometric_tensor())`.
```

The 3 implemnentations are:

 - {ref}`netket.optimizer.qgt.QGTOnTheFly`, which uses jax automatic differentiation through 2 vjp and 1 jvp product to compute the action of quantum geometric tensor on a vector and operates natively on PyTrees. This method shines if the parameters of your network are stored in a PyTree with few leaf nodes.
 - {ref}`netket.optimizer.qgt.QGTJacobianDense`, which precomputes the log derivatives ( $ O_k $ ) when it's constructed and converts it to a single dense array. If you have a high number of total parameters or many leaf nodes in your parameter PyTree, this implementation might perform better because everything is stored contiguously in memory. However, it has an high 'startup cost'. This method might not work and/or give wrong results if you mix real and complex parameters.
 - {ref}`netket.optimizer.qgt.QGTJacobianPyTree`, same as above, but the precomputed jacobian is not stored contiguously in memory but is stored as a PyTree. This will work better than `QGTJacobianDense` if there are few leaf nodes, but better than `QGTOnTheFly` if you use iterative solvers that perform many steps.

> We also have an extra implemnetation, called `netket.optimizer.qgt.QGTAuto`, which uses some heuristics based on the parameters of the network to select the best QGT implementation. Be warned that the heuristics we use is very crude, and might not pick the best implementation all the time.

Reviewer: ```suggestion
We also have an extra implementation, called `netket.optimizer.qgt.QGTAuto`, which uses some heuristics based on the parameters of the network to select the best QGT implementation. Be warned that the heuristics we use is very crude, and might not pick the best implementation all the time.
```

All the QGT implemnetation listed above have several options that can affect their performance. 
We advise you to have a look at them and experiment.
We provide a lot of freedom because it's not yet clear to us what is the best implementation for which kind of problems.
If you work with NetKet and determine that a particular implementation works best for certain types of networks, we would be glad to hear it! Let us know by opening a Discussion on our GitHub repository. We might use the insight to improve the automatic selection.

## Stochastic Reconfiguration

The [stochastic reconfiguration method](https://www.attaccalite.com/PhDThesis/html/node15.html) is a technique that makes use of the information encoded in the quantum geometric tensor described above to _precondition_ the gradient used in stochastic optimization, _improving_ the convergence rate to the ground state.

> We would like to underline that Stochastic Reconfiguration can be derived (and therefore though as) an imaginary time evolution. 

Reviewer 2: ```suggestion
We would like to underline that Stochastic Reconfiguration can be derived (and therefore thought as) an imaginary time evolution. 
```
